Sources/SwiftCoAP/SCMessage.swift

```diff
@@ -32,87 +34,171 @@ protocol SCCoAPTransportLayerDelegate: class {
 //MARK:
 //MARK: SC CoAP Transport Layer Protocol declaration
 
-protocol SCCoAPTransportLayerProtocol: class {
+public protocol SCCoAPTransportLayerProtocol: AnyObject {
     //SCClient uses this property to assign itself as delegate
     var transportLayerDelegate: SCCoAPTransportLayerDelegate! { get set }
     
     //SClient calls this method when it wants to send CoAP data
     func sendCoAPData(_ data: Data, toHost host: String, port: UInt16) throws
-    
+
     //Called when the transmission is over. Clear your states (e.g. close sockets)
     func closeTransmission()
     
     //Start to listen for Messages. Prepare e.g. sockets for receiving data. This method will only be called by SCServer
     func startListening() throws
 }
 
+struct HostPortKey: Hashable {
+    var host: String
+    var port: UInt16
+}
+
 
 
 //MARK:
 //MARK: SC CoAP UDP Transport Layer: This class is the default transport layer handler, sending data via UDP with help of GCDAsyncUdpSocket. If you want to create a custom transport layer handler, you have to create a custom class and adopt the SCCoAPTransportLayerProtocol. Next you have to pass your class to the init method of SCClient: init(delegate: SCClientDelegate?, transportLayerObject: SCCoAPTransportLayerProtocol). You will than get callbacks to send CoAP data and have to inform your delegate (in this case an object of type SCClient) when you receive a response by using the callbacks from SCCoAPTransportLayerDelegate.
 
-final class SCCoAPUDPTransportLayer: NSObject {
-    weak var transportLayerDelegate: SCCoAPTransportLayerDelegate!
-    var udpSocket: GCDAsyncUdpSocket!
-    var port: UInt16 = 0
-    fileprivate var udpSocketTag: Int = 0
-    
-    convenience init(port: UInt16) {
-        self.init()
-        self.port = port
+public final class SCCoAPUDPTransportLayer: NSObject {
+    weak public var transportLayerDelegate: SCCoAPTransportLayerDelegate!
+    var connections: [HostPortKey: NWConnection] = [:]
+    var listener: NWListener?
+
+    private func setupStateUpdateHandler(for connection: NWConnection, withHostPort hostPort: HostPortKey) -> NWConnection {
+        connection.stateUpdateHandler = { [weak self] newState in
+            switch newState {
+            case .failed(let error):
+                os_log("Connection FAILED", log: .default, type: .error, "\(error)")
+                self?.connections[hostPort]?.cancel()
+                self?.connections.removeValue(forKey: hostPort)
+                guard let self = self else { return }
+                self.transportLayerDelegate?.transportLayerObject(self, didFailWithError: error as NSError)
```

Comment:
Why the `transportLayerDelegate?` if it's force-unwrapped in the definition? 

Maybe switch the `public var transportLayerDelegate` to an optional to avoid possible crashes? Since the delegate is not required.